doc: Add limitations of CL IDs in URLS for perforce depots (#54249)
sourcegraph/sourcegraph: doc/admin/repo/perforce.md

- Viewing a specific CL
- Viewing the files of a depot at a specific CL
- Viewing a specific file added / removed / modified in a specific CL
- Viewing the list of CLs

#### Limitations

While we are working to bring this feature to all areas of the Sourcegraph search experience in subsequent releases there are a few limitations that should be aware of while using this feature:

- To support CLs in the URLs natively, Sourcegraph needs to perform background computation post clone / fetch of a depot by parsing each commit to retrieve the CL ID and store it in the `repo_commits_changelists` table
- This background computation currently is performed on only one depot at a time and we are working to support this for multiple depots in parallel
- Although it doesn't take hours to complete the mapping of commits to SHAs, it can take upto a few minutes the first time a new depot is added or this experimental configuration is enabled
- When a background mapping job is running - the depot won't be actively serviceable as URLs to CL IDs may not resolve and users may see an error while interacting with the repo

> - As a result it's recommended that site admins use a maintenance window to enable this flag especially if they have a large number of Perforce depots or one or more large depots with hundreds of thousands of CLs

Reviewer: Again: doesn't read like a separate bullet point

> While we are working to bring this feature to all areas of the Sourcegraph search experience in subsequent releases there are a few limitations that should be aware of while using this feature:
> 
> - To support CLs in the URLs natively, Sourcegraph needs to perform background computation post clone / fetch of a depot by parsing each commit to retrieve the CL ID and store it in the `repo_commits_changelists` table
> - This background computation currently is performed on only one depot at a time and we are working to support this for multiple depots in parallel
> - Although it doesn't take hours to complete the mapping of commits to SHAs, it can take upto a few minutes the first time a new depot is added or this experimental configuration is enabled
> - When a background mapping job is running - the depot won't be actively serviceable as URLs to CL IDs may not resolve and users may see an error while interacting with the repo
> - As a result it's recommended that site admins use a maintenance window to enable this flag especially if they have a large number of Perforce depots or one or more large depots with hundreds of thousands of CLs
> - While removing a depot from a code host config will mark it as "deleted", the mapped information will **not** be deleted to prevent forced recomputation after an accidental removal of depot from a code host config
> - This experimental configuration can not be selectively enabled for a specific perforce depot

Reviewer 2: ```suggestion
- After a depot clone/fetch, Sourcegraph computes and stores mappings of changelist IDs to commit hashes. This mapping can take several minutes for large clones/fetches. While the mapping is ongoing, new changelist IDs  will not be available for use in URLs.
```

- Since you started with a `Limitations` heading, you don't need the intro comment. Just a straightforward list of gotchas is perfectly fine.
- as @mrnugget pointed out, the bullet points can be combined, and the language about how long it takes can be simplified.
- I removed the mention of the table name (`repo_commits_changelists`), but since this documentation is for admins, it may be appropriate to include it. I'm not sure how deep our admins dive.
- The discussion about the "soft deleting" is good information for admins, but it is not a "Limitation". It may fit better in another section. Maybe add a "Deep dive" section and add the information about the table name along with the information about the "soft delete"
- I'm not sure about the maintenance window suggestion. If the feature is being enabled for the first time, there will be no CID URLs in use. If a large clone/fetch happens after the feature is enabled, the CID URLs will also be new. If there is a massive re-clone event after having the feature enabled for awhile, and the users regularly use CID URLs, then maybe a maintenance window would be appropriate, but in that case, they'll probably already have a maintenance window in place because of the large re-clone.

Author: > I removed the mention of the table name (repo_commits_changelists), but since this documentation is for admins, it may be appropriate to include it. I'm not sure how deep our admins dive.

I would keep it around because the doc is also referenced by CE / CS when they are troubleshooting for customers. I'll reorganise it in another section along with the soft deleting information.

> I'm not sure about the maintenance window suggestion. If the feature is being enabled for the first time, there will be no CID URLs in use. If a large clone/fetch happens after the feature is enabled, the CID URLs will also be new. If there is a massive re-clone event after having the feature enabled for awhile, and the users regularly use CID URLs, then maybe a maintenance window would be appropriate, but in that case, they'll probably already have a maintenance window in place because of the large re-clone.

For existing depots already cloned to Sourcegraph we will need a maintenance window and it may not be obvious to the site admin if they're just setting a site config.

As of Sourcegraph 5.1, there are new features for Perforce depots that need to be enabled.

### Changelist ID in URLs

> Note: It's recommended that site admins use a maintenance window to enable the `perforceChangelistMapping` flag especially if they have a large number of Perforce depots or one or more large depots with hundreds of thousands of changelists (see limitations below).

Reviewer: Why not use the `> NOTE: ` syntax here?

Also: the note is missing _why_. As someone who reads this the first time, I'd be scared off immediately. Why should I use a maintenance window? will the instance go down? be unreachable? I'd at least mention something like "use a maintenance window, since the perforce depots in sourcegraph will be unavailable for a few minutes".

Author: Should I move this back into the limitations section? The why is part of the limitations but I get why it can be scray / odd for a user reading this for the first time.

Reviewer: No, I think this can be reformulated.

Example:

```suggestion
Note: When enabling changelist IDs in URLs for the first time, Perforce depots can be unavailable for a few minutes on the Sourcegraph instance, due to the initial mapping of changelist IDs to generated commit ID happening in the background. If you have a large number of Perforce depots, we recommend proceeding with the following steps in a maintenance window in which you don't expect large amounts of traffic on your Sourcegraph instance.
```

Author: Wow this is so much better! Thank you so much! I'm going to apply this as is.

Add `"perforceChangelistMapping": "enabled",` to `experimentalFeatures` in the [site configuration](../config/site_config.md):

```json
{
  "experimentalFeatures": {
    "perforceChangelistMapping": "enabled"
  }
}
```

When enabled, URLs for Perforce code hosts will use the Changelist (CL) ID instead of commit SHAs. Areas that benefit from this at the moment are:

- Viewing a specific CL
- Viewing the files of a depot at a specific CL
- Viewing a specific file added / removed / modified in a specific CL
- Viewing the list of CLs

#### Limitations

- After a depot is cloned or fetched, Sourcegraph computes and stores mappings of CL IDs to commit SHAs. This mapping can take several minutes for large clones/fetches. When a background mapping job is running, the depot won't be serviceable as URLs referring to CL IDs may not resolve and users may see an error while interacting with the depot.
- This experimental configuration can not be selectively enabled for a specific perforce depot.

#### Mechanism

> To support CLs natively in the URLs, Sourcegraph performs background computation post clone / fetch of a depot by parsing each commit to retrieve the CL ID and store it in the `repo_commits_changelists` table. This is currently performed on only one depot at a time and we are working to support this for multiple depots in parallel in an upcoming release.

Reviewer: ```suggestion
To support CLs natively in the URLs, Sourcegraph performs background computation after syncing the contents of a depot. That's done by parsing each generated commit to retrieve the corresponding CL ID and store it in the `repo_commits_changelists` table. This is currently performed on only one depot at a time and we are working to support this for multiple depots in parallel in an upcoming release.
```
